From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AJMFactsheets <AJMFactsheets@gmail.com>
Date: Wed, 22 Jan 2020 19:52:28 -0600
Subject: [PATCH] Fix items vanishing through end portal

If the Paper configuration option "keep-spawn-loaded" is set to false,
items entering the overworld from the end will spawn at Y = 0.

This is due to logic in the getHighestBlockYAt method in World.java
only searching the heightmap if the chunk is loaded.

Quickly loading the exact world spawn chunk before searching the
heightmap resolves the issue without having to load all spawn chunks.

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index d8cb7a53940c244b3ad36d32cdbbee1b3746e543..6b11da94839a677c6bf63fd2d8d4994754e3ad05 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3210,6 +3210,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
             if (flag1) {
                 blockposition1 = ServerLevel.END_SPAWN_POINT;
             } else {
+                // Paper start - Ensure spawn chunk is always loaded before calculating Y coordinate
+                destination.getChunkAt(destination.getSharedSpawnPos());
+                // Paper end
                 blockposition1 = destination.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, destination.getSharedSpawnPos());
             }
             // CraftBukkit start
